### DISCUSSION TITLE
Querying vehicle_data (or other) from API

### DISCUSSION THREAD

**Discussion by clarity-phev** [teslamate-org/teslamate#4449] [Q&A]
Does teslamate include any functionality to perform manual queries or send commands through the API for diagnostic / debugging purposes?  Since it already has already gone through the authentication and has the appropriate token, etc. it would seem like an easy function to provide.

Trying to do this directly through the API seems like a daunting task for a non software developer...

**Comment by cwanja** [comment]
I assume you mean sending a command to the Tesla API and not a TeslaMate API?

If so, we could introduce that feature. But it's not there today. Enhanced security would need to be done and that comes in the form of more experience needed to configure and run TeslaMate. 

With the API changes coming (e.g. charging) that could be a roadblock as well.

**Reply by clarity-phev** [reply]
Yes, I was referring to access directly to/from the Tesla API.
Thanks for considering this.  You mention enhanced security, more complicated configuration...  Once TeslaMate has already gained access to the Tesla API could it interact back and forth via a 'simple' console window?  I know threats to the unofficial API loom over all of this...

Do you have a suggestion as to how to the best do a simple command or query of the Tesla API (tools, or complete examples) without getting mired in a lot of tricky programming?  I naively assumed it would be a couple of 'curls', but it has become clear to me that it is more involved.

**Reply by cwanja** [reply]
Believe commands are blocked behind the Fleet API which is your first barrier to overcome. Tesla has all of that documented and we have some here. https://docs.teslamate.org/docs/guides/api#guide-to-using-the-official-tesla-api-directly-free

This was the case a few months ago last time I tried a few CURL commands. 

What we believe today is that everyone would have to do those setups to migrate to the Fleet API (or just send commands like you want). And as you can see, the requirements:
- Advanced IT skills
- A dedicated public hosting
- A free Tesla Fleet API plan

**Reply by clarity-phev** [reply]
What actually led me down this rabbit hole was your post here:
https://github.com/teslamate-org/teslamate/issues/4368#issuecomment-2466831807

How did @VintageWillis perform the query that you requested him to do ('vehicle_config' -> 'car_type')?

**Reply by cwanja** [reply]
Few factors. That's a simple query to _get_ data, not perform a command. Additionally, the Fleet API only supports certain vehicles and MCUs.

Hope this is becoming clearer why we did not migrate to Fleet API nor include commands. There is healthy discussion in #3792 and #4408

**Reply by clarity-phev** [reply]
Thanks for the clarification...
Let's say that I am happy being able to do a 'simple query to get data'.  Can you elaborate on how to do that with the current configuration of TeslaMate (non-Fleet API)?
I don't have a specific agenda or need for commands, but am intrigued about how to query some additional information.

**Reply by cwanja** [reply]
You need to get your bearer token and then construct a CURL command against the API endpoint with the appropriate parameters. https://tesla-api.timdorr.com/ used to be the go-to guide, but I am not sure if it is still accurate.

**Reply by clarity-phev** [reply]
I have working tokens since I've been happily using TeslaMate for over a year now.  Constructing a CURL command against the API endpoint is the elusive part to me.  https://tesla-api.timdorr.com/ is an excellent resource, but it is short on details / examples as to specifically how to construct a valid command which uses the endpoints.

I will keep plugging along, and the formula will hopefully become clear.

Thanks for the pointers, and for all of the work that you do to keep TeslaMate alive !
It is an excellent program and I continue to hope that Tesla doesn't totally abandon their technical owners.

**Comment by arcastro** [comment]
Teslamate publishes a lot of data it receives from the tesla API to MQTT. It parses the data from the API and publishes it to individual topics that you can subscribe to to receive events when the data for that topic changes.

If you already have teslamate running, you can `docker exec` into the existing MQTT container and subscribe to all topics to see some of the data available by running `mosquitto_sub -h localhost -t '#' -v`

You'll see something like this:

```
/ # mosquitto_sub -h localhost -t '#' -v
teslamate/cars/1/odometer 15406.63
teslamate/cars/1/tpms_soft_warning_fr false
teslamate/cars/1/tpms_soft_warning_rl false
teslamate/cars/1/tpms_soft_warning_rr false
teslamate/cars/1/plugged_in true
teslamate/cars/1/latitude 123.45678
teslamate/cars/1/outside_temp 13.5
teslamate/cars/1/exterior_color StealthGrey
teslamate/cars/1/doors_open false
teslamate/cars/1/is_preconditioning false
teslamate/cars/1/longitude 123.45678
teslamate/cars/1/locked true
teslamate/cars/1/active_route_latitude nil
...etc

(note: I redacted latitude and longitude above for privacy)
```

Data is periodically polled/streamed, so as it gets updated and published to each topic, you will see a new log line for that topic and its corresponding most recent data.

Of course, you probably have a programmatic use-case in mind. You can use any MQTT library for your programming language of choice and subscribe to whatever topics you are interested in receiving data from. For example, if your application subscribes to the `teslamate/cars/1/latitude` topic, then your application will receive events whenever the latitude changes.

Since much of the data is already in MQTT, you can integrate with that instead of integrating with the tesla api derectly, depending on your use-case. However, this might break at any time if tesla decides to limit access to existing owner/streaming api #4408

Not sure about sending commands. This piece is for read-only data.

**Reply by arcastro** [reply]
> for a non software developer...

Ah apologies. Somehow I missed this part when writing this comment 😅. Feel free to disregard.
I'll leave this comment up in case it helps anyone else though. And you might still be interested in taking a peek at some of the data in MQTT.

**Reply by clarity-phev** [reply]
Thanks for the thoughtful response...

I am aware of the MQTT capabilities of TeslaMate.  In fact, I have been running tesla-geogdo which uses MQTT to automate the garage door.  I also have been using a tool called 'pushover' to monitor MQTT messages and send notifications to our cell phones when there are garage door actions.

For debugging, I like a tool called 'MQTTX' (from a Linux laptop) to noodle around with the MQTT messaging particularly when working to get 'tesla-geogdo' running.

Although I am not a 'software developer', I do have 35 years of engineering experience including a fair amount of embedded programming so I do greatly appreciate your inputs !

I don't really have a specific use case, but rather, was simply intrigued by the scope of the unofficial owner's API and I was thinking that TeslaMate may only use a subset of the available API  capabilities.  I was just wondering how to explore everything that is available (and better understand the API itself).  Perhaps TeslaMate makes more extensive use of the API than I was thinking.

Like everyone here, I am dreading what seems to be inevitable (Tesla shutting down the unofficial API), and I hope this sort of owner access to technical telemetry can continue.  It is likely to be crippled by limited numbers of calls, etc., but I hope it remains largely viable.  Tesla should never abandon it's technical owner community.  It is a large part of why Tesla has become so successful.

**Comment by clarity-phev** [comment]
As a partial answer to my own question, I found a way (with a simple https: call) to query some information about the vehicle.   All that is required is a valid 'refresh token' which any TeslaMate user is going to have.

Here is the URL that works:
https://tesla-info.com/api/control_v3.php?refresh=EXTREMELY_LONG_REFRESH_TOKEN_GOES_HERE&request=get_data

It returns with the following JSON message (sanitized with 'X's' where personal data goes)...

```	"car": "",
	"status": "valid",
	"Response": {
		"response": {
			"id": XXXXXXXXXXXXXXXX,
			"user_id": XXXXXXXXXXXXXXXX,
			"vehicle_id": XXXXXXXXXXXXXXXXXXX,
			"vin": "7SAYXXXXXXXXXXXX",
			"color": null,
			"access_type": "OWNER",
			"granular_access": {
				"hide_private": false
			},
			"tokens": null,
			"state": "online",
			"in_service": false,
			"id_s": "XXXXXXXXXXXXXXXX",
			"calendar_enabled": true,
			"api_version": 84,
			"backseat_token": null,
			"backseat_token_updated_at": null,
			"ble_autopair_enrolled": false,
			"supercharger_payment_needed": "false",
			"supercharging_enabled": true,
			"charge_state": {
				"battery_heater_on": false,
				"battery_level": 67,
				"battery_range": 186.28,
				"charge_amps": 48,
				"charge_current_request": 48,
				"charge_current_request_max": 48,
				"charge_enable_request": true,
				"charge_energy_added": 22.04,
				"charge_limit_soc": 80,
				"charge_limit_soc_max": 100,
				"charge_limit_soc_min": 50,
				"charge_limit_soc_std": 80,
				"charge_miles_added_ideal": 91,
				"charge_miles_added_rated": 91,
				"charge_port_cold_weather_mode": false,
				"charge_port_color": "Off",
				"charge_port_door_open": false,
				"charge_port_latch": "Engaged",
				"charge_rate": 0,
				"charger_actual_current": 0,
				"charger_phases": null,
				"charger_pilot_current": 48,
				"charger_power": 0,
				"charger_voltage": 2,
				"charging_state": "Disconnected",
				"conn_charge_cable": "<invalid>",
				"est_battery_range": 210.92,
				"fast_charger_brand": "<invalid>",
				"fast_charger_present": false,
				"fast_charger_type": "<invalid>",
				"ideal_battery_range": 186.28,
				"max_range_charge_counter": 0,
				"minutes_to_full_charge": 0,
				"not_enough_power_to_heat": null,
				"off_peak_charging_enabled": false,
				"off_peak_charging_times": "all_week",
				"preconditioning_enabled": false,
				"preconditioning_times": "all_week",
				"scheduled_charging_mode": "Off",
				"scheduled_charging_pending": false,
				"scheduled_charging_start_time": null,
				"scheduled_departure_time": null,
				"supercharger_session_trip_planner": false,
				"time_to_full_charge": 0,
				"timestamp": 1734880972579,
				"trip_charging": false,
				"usable_battery_level": 67,
				"user_charge_enable_request": null
			},
			"climate_state": {
				"allow_cabin_overheat_protection": true,
				"auto_seat_climate_left": false,
				"auto_seat_climate_right": false,
				"auto_steering_wheel_heat": false,
				"battery_heater": false,
				"battery_heater_no_power": null,
				"bioweapon_mode": false,
				"cabin_overheat_protection": "Off",
				"cabin_overheat_protection_actively_cooling": false,
				"climate_keeper_mode": "off",
				"cop_activation_temperature": "High",
				"defrost_mode": 0,
				"driver_temp_setting": 21,
				"fan_status": 0,
				"hvac_auto_request": "On",
				"inside_temp": 9,
				"is_auto_conditioning_on": false,
				"is_climate_on": false,
				"is_front_defroster_on": false,
				"is_preconditioning": false,
				"is_rear_defroster_on": false,
				"left_temp_direction": 0,
				"max_avail_temp": 28,
				"min_avail_temp": 15,
				"outside_temp": -4.5,
				"passenger_temp_setting": 21,
				"remote_heater_control_enabled": false,
				"right_temp_direction": 0,
				"seat_heater_left": 0,
				"seat_heater_rear_center": 0,
				"seat_heater_rear_left": 0,
				"seat_heater_rear_right": 0,
				"seat_heater_right": 0,
				"side_mirror_heaters": false,
				"steering_wheel_heat_level": 0,
				"steering_wheel_heater": false,
				"supports_fan_only_cabin_overheat_protection": true,
				"timestamp": 1734880972579,
				"wiper_blade_heater": false
			},
			"drive_state": {
				"power": 0,
				"shift_state": null,
				"speed": null,
				"timestamp": 1734880972579
			},
			"gui_settings": {
				"gui_24_hour_time": false,
				"gui_charge_rate_units": "kW",
				"gui_distance_units": "mi/hr",
				"gui_range_display": "Rated",
				"gui_temperature_units": "F",
				"gui_tirepressure_units": "Psi",
				"show_range_units": false,
				"timestamp": 1734880972579
			},
			"vehicle_config": {
				"aux_park_lamps": "NaPremium",
				"badge_version": 1,
				"can_accept_navigation_requests": true,
				"can_actuate_trunks": true,
				"car_special_type": "base",
				"car_type": "modely",
				"charge_port_type": "US",
				"cop_user_set_temp_supported": true,
				"dashcam_clip_save_supported": true,
				"default_charge_to_max": false,
				"driver_assist": "TeslaAP4",
				"ece_restrictions": false,
				"efficiency_package": "MYSRPlus2022Q1TX",
				"eu_vehicle": false,
				"exterior_color": "PearlWhite",
				"exterior_trim": "Black",
				"exterior_trim_override": "",
				"has_air_suspension": false,
				"has_ludicrous_mode": false,
				"has_seat_cooling": false,
				"headlamp_type": "Premium",
				"interior_trim_type": "White2",
				"key_version": 2,
				"motorized_charge_port": true,
				"paint_color_override": "",
				"performance_package": "Base",
				"plg": true,
				"pws": true,
				"rear_drive_unit": "PM216MOSFET",
				"rear_seat_heaters": 1,
				"rear_seat_type": 0,
				"rhd": false,
				"roof_color": "RoofColorGlass",
				"seat_type": null,
				"sentry_preview_supported": true,
				"spoiler_type": "None",
				"sun_roof_installed": null,
				"supports_qr_pairing": false,
				"third_row_seats": "None",
				"timestamp": 1734880972579,
				"trim_badging": "50d",
				"use_range_badging": true,
				"utc_offset": -18000,
				"webcam_selfie_supported": true,
				"webcam_supported": true,
				"wheel_type": "Apollo19"
			},
			"vehicle_state": {
				"api_version": 84,
				"autopark_state_v3": "unavailable",
				"calendar_supported": true,
				"car_version": "2024.44.25.2 288ce84bb43e",
				"center_display_state": 0,
				"dashcam_clip_save_available": true,
				"dashcam_state": "Recording",
				"df": 0,
				"dr": 0,
				"fd_window": 0,
				"feature_bitmask": "fbdffbff,ae5cbc7f,5",
				"fp_window": 0,
				"ft": 0,
				"is_user_present": false,
				"locked": true,
				"media_info": {
					"audio_volume": 4,
					"audio_volume_increment": 0.333333,
					"audio_volume_max": 10.333333,
					"media_playback_status": "Playing",
					"now_playing_album": "",
					"now_playing_artist": "",
					"now_playing_duration": -1000,
					"now_playing_elapsed": 0,
					"now_playing_source": "TuneInWeb",
					"now_playing_station": "",
					"now_playing_title": ""
				},
				"media_state": {
					"remote_control_enabled": true
				},
				"notifications_supported": true,
				"odometer": 16955.016638,
				"parsed_calendar_supported": true,
				"pf": 0,
				"pr": 0,
				"rd_window": 0,
				"remote_start": false,
				"remote_start_enabled": true,
				"remote_start_supported": true,
				"rp_window": 0,
				"rt": 0,
				"santa_mode": 0,
				"sentry_mode": false,
				"sentry_mode_available": true,
				"service_mode": false,
				"service_mode_plus": false,
				"software_update": {
					"download_perc": 0,
					"expected_duration_sec": 2700,
					"install_perc": 1,
					"status": "",
					"version": " "
				},
				"speed_limit_mode": {
					"active": false,
					"current_limit_mph": 85,
					"max_limit_mph": 120,
					"min_limit_mph": 50,
					"pin_code_set": false
				},
				"timestamp": 1734880972579,
				"tpms_hard_warning_fl": false,
				"tpms_hard_warning_fr": false,
				"tpms_hard_warning_rl": false,
				"tpms_hard_warning_rr": false,
				"tpms_last_seen_pressure_time_fl": 1734879817,
				"tpms_last_seen_pressure_time_fr": 1734879846,
				"tpms_last_seen_pressure_time_rl": 1734879846,
				"tpms_last_seen_pressure_time_rr": 1734879817,
				"tpms_pressure_fl": 2.725,
				"tpms_pressure_fr": 2.725,
				"tpms_pressure_rl": 2.7,
				"tpms_pressure_rr": 2.7,
				"tpms_rcp_front_value": 2.9,
				"tpms_rcp_rear_value": 2.9,
				"tpms_soft_warning_fl": false,
				"tpms_soft_warning_fr": false,
				"tpms_soft_warning_rl": false,
				"tpms_soft_warning_rr": false,
				"valet_mode": false,
				"valet_pin_needed": true,
				"vehicle_name": null,
				"vehicle_self_test_progress": 0,
				"vehicle_self_test_requested": false,
				"webcam_available": true
			}
		}
	}
}
```

I just found this construct as an example somewhere online, but what is still totally unclear is how to construct a similar query for any arbitrary parameter that is documented in the undocumented API.  Nor am I clear as to whether this query is one of the ones in the undocumented API (or which it is) !!

Perhaps this query is one that is done by TeslaMate and all of this is available from TeslaMate as MQTT ?  Still very muddy to me !!
